src/main/java/com/google/devtools/build/lib/bazel/bzlmod/GitRepoSpecBuilder.java

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.devtools.build.lib.bazel.bzlmod;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import net.starlark.java.eval.StarlarkInt;
+import java.util.List;
+
+/**
+ * Builder for a {@link RepoSpec} object that indicates how to materialize a repo corresponding to
+ * a {@code git_repository} repo rule call.
+ */
+public class GitRepoSpecBuilder {
+
+  public static final String GIT_REPO_PATH = "@bazel_tools//tools/build_defs/repo:git.bzl";
+
+  private final ImmutableMap.Builder<String, Object> attrBuilder;
+
+  public GitRepoSpecBuilder() {
+    attrBuilder = new ImmutableMap.Builder<>();
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setRepoName(String repoName) {
+    return setAttr("name", repoName);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setRemote(String remoteRepoUrl) {
+    return setAttr("remote", remoteRepoUrl);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setCommit(String gitCommitHash) {
+    return setAttr("commit", gitCommitHash);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setShallowSince(String shallowSince) {
+    return setAttr("shallow_since", shallowSince);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setTag(String tag) {
+    return setAttr("tag", tag);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setBranch(String branch) {
+    return setAttr("branch", branch);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setInitSubmodules(boolean initSubmodules) {
+    return setAttr("init_submodules", initSubmodules);
+  }
+
+  @CanIgnoreReturnValue
+  public GitRepoSpecBuilder setRecursiveInitSubmodules(boolean recursiveInitSubmodules) {
+    return setAttr("recursive_init_submodules", recursiveInitSubmodules);
```

Comment:
i just submitted https://github.com/bazelbuild/bazel/pull/20937 where i collapsed this field with the other one, should we do the same here? (or should i change mine to not do that?) @Wyverald wdyt?

Reply:
what is the historical reason for the difference between the two? Should we just deprecate the "recursive" one?

Judging by https://cs.opensource.google/bazel/bazel/+/master:tools/build_defs/repo/git_worker.bzl;drc=c2553cc6ccece9826895bd691c4981d1e790e9e6;l=167, is it because everyone has a newer Git installed now, so `--recursive` is no longer necessary ever?

Reply:
I added the recursive one and I think i didn't flip that default to just not break behavior https://github.com/bazelbuild/bazel/pull/10545

Reply:
I think you still need to pass `--recursive`

Reply:
> I added the recursive one and I think i didn't flip that default to just not break behavior #10545

With #20937 I assumed you meant that these fields were only separate due to historical reasons. Did I misunderstand then? Why does #20937 collapse the two fields if we still need to pass `--recursive` sometimes?

Reply:
It chooses to always pass recursive if you need submodules at all instead. Which I think is probably the more common case. Theoretically you could have direct submodules were you don't care to clone their transitive submodules but I assume that is rare

Reply:
oh I see -- the two are collapsed as in both are set to true. I don't have an opinion wrt whether to collapse or not (no experience with git submodules), but I do think we should aim to make our APIs consistent.

Reply:
Is there any AI on this change?

Reply:
https://github.com/bazelbuild/bazel/commit/5cf7714faffeaaa3ca017e526d232af205f5fb85 ended up being submitted with the "collapse" option, so let's go with that. @nachumg, would you mind making GitOverride use GitRepoSpecBuilder, and removing the `recursiveInitSubmodules` field from source.json and instead use the `initSubmodules` field for both `init_submodules` and `recursive_init_submodules` attributes in GitRepoSpecBuilder?

Reply:
Done. But note this means I re-added the patch properties to the json source.

Reply:
Sorry to keep nagging about it, but you don't have to add the patch properties back to source.json; GitRepoSpecBuilder can still have the patch methods, but they're never called by IndexRegistry, only by GitOverride. I feel somewhat strongly that we shouldn't expose the patch fields in source.json.